tests/rptest/services/kgo_verifier_services.py

```diff
@@ -53,91 +52,236 @@ def __init__(self, context, redpanda, topic, msg_size, custom_node):
+            return super(KgoVerifierService, self).allocate_nodes()
 
     def free(self):
         if self.use_custom_node:
             return
         else:
-            return super(KgoVerifierServiceService, self).free()
+            return super(KgoVerifierService, self).free()
 
 
-class ConsumerStatus:
-    def __init__(self, valid_reads, invalid_reads, out_of_scope_invalid_reads):
+class StatusThread(threading.Thread):
+    INTERVAL = 5
+
+    def __init__(self, parent: Service, node, status_cls, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self._parent = parent
+        self._node = node
+        self._status_cls = status_cls
+        self._last_read = threading.Event()
+        self._stop_now = threading.Event()
+        self._ex = None
+        self._ready = False
+
+    @property
+    def errored(self):
+        return self._ex is not None
+
+    @property
+    def who_am_i(self):
+        return self._parent.who_am_i()
+
+    @property
+    def logger(self):
+        return self._parent.logger
+
+    def raise_on_error(self):
+        if self._ex is not None:
+            raise self._ex
+
+    def run(self):
+        try:
+            # Don't assume the HTTP port will be open instantly on startup (although it should be open very soon)
+            self.await_ready()
+
+            self.poll_status()
+        except Exception as ex:
+            self._ex = ex
+            self.logger.exception(
+                f"Error reading status from {self.who_am_i} on {self._node.name}"
+            )
+
+    def is_ready(self):
+        try:
+            r = requests.get(self._parent._remote_url(self._node, "status"))
+        except (requests.exceptions.RequestException,
+                requests.exceptions.ConnectionError) as e:
+            self.logger.debug(
+                f"Status endpoint {self.who_am_i} not ready: {e}")
+            return False
+        except Exception as e:
+            self.logger.debug(
+                f"Status endpoint {self.who_am_i} not ready: {e}")
+            return False
+        else:
+            return r.status_code == 200
+
+    def _ingest_status(self, worker_statuses):
+        reduced = self._status_cls(**worker_statuses[0])
+        for s in worker_statuses[1:]:
+            reduced.merge(self._status_cls(**s))
+
+        self._parent._status = reduced
+
+    def await_ready(self):
+        """
+        Wait for the remote processes http endpoint to come up
+        """
+        if self._ready:
+            return
+
+        wait_until(
+            lambda: self.is_ready() or self._stop_now.is_set(),
+            timeout_sec=5,
+            backoff_sec=0.5,
+            err_msg=
+            f"Timed out waiting for status endpoint {self.who_am_i} to be available"
+        )
+        self._ready = True
+
+    def poll_status(self):
+        while not self._stop_now.is_set():
+            drop_out = self._last_read.is_set()
```

Comment:
I'll add a comment.  This distinction is that stop_now means quit, and _last_read means read status once more and then quit.

Reply:
Right, I understand that, but I'm unsure the distinction of when `shutdown()` is preferred over `stop()`, given they both wait for at least the current round to finish (and in `shutdown()`'s case, one more). I guess in a general sense it's for tests that perform some action, and then shut down the service, expecting to see a side effect from that action in the last status.

Reply:
When wait()'ing on a service, we first allow the worker to finish its loop, then tear down the status thread.

If we used stop(), then we would see the last status we happened to query before that point, which might not be after the worker finished its loop: if the worker had encountered a validation error on the last message it read, we might not see it.

When we use shutdown(), that extra read of the status ensures that any issues found between our call to the service's "/last_pass" and the present will have been read.

So for the service's wait() we want the thread's shutdown(), so that we definitely see all the worker's status up to its finish, but for the service's stop() we don't care, we just want to drop out.

Reply:
Ah, the dot I wasn't connecting was the fact that `shutdown` is paired up with a `/last_pass` request, and that may have side effects that we want to capture in observing the last status. Thanks for explaining!